core/src/main/java/org/infinispan/persistence/util/PersistenceManagerCloseableSupplier.java

```diff
@@ -0,0 +1,159 @@
+   private final Executor executor;
+   private final PersistenceManager manager;
+   private final KeyFilter<K> filter;
+   private final InternalEntryFactory factory;
+   private final BlockingQueue<CacheEntry<K, V>> queue;
+   private final long timeout;
+   private final TimeUnit unit;
+
+   private final Lock closeLock = new ReentrantLock();
+   private final Condition closeCondition = closeLock.newCondition();
+
+   private boolean closed = false;
+   private AtomicReference<AdvancedCacheLoader.CacheLoaderTask<K, V>> taskRef = new AtomicReference<>();
+
+   public PersistenceManagerCloseableSupplier(Executor executor, PersistenceManager manager,
+                                              InternalEntryFactory factory, KeyFilter<K> filter, long timeout,
+                                              TimeUnit unit, int maxQueue) {
+      this.executor = executor;
+      this.manager = manager;
+      this.factory = factory;
+      this.filter = filter;
+      this.timeout = timeout;
+      this.unit = unit;
+      this.queue = new ArrayBlockingQueue<>(maxQueue);
+   }
+
+   class SupplierCacheLoaderTask implements AdvancedCacheLoader.CacheLoaderTask<K, V> {
+
+      @Override
+      public void processEntry(MarshalledEntry<K, V> marshalledEntry, AdvancedCacheLoader.TaskContext taskContext)
+              throws InterruptedException {
+         if (!taskContext.isStopped()) {
+            closeLock.lock();
+            try {
+               if (closed) {
+                  taskContext.stop();
+                  return;
+               }
+            } finally {
+               closeLock.unlock();
+            }
+            InternalCacheEntry<K, V> ice = PersistenceUtil.convert(marshalledEntry, factory);
+            // We do a read without acquiring lock
+            boolean stop = closed;
+            while (!stop) {
+               // If we were able to offer a value this means someone took from the queue so let us come back around main
+               // loop to offer all values we can
+               // TODO: do some sort of batching here - to reduce wakeup contention ?
+               if (queue.offer(ice, 100, TimeUnit.MILLISECONDS)) {
+                  closeLock.lock();
+                  try {
+                     // Wake up anyone waiting for a value
+                     closeCondition.signalAll();
+                  } finally {
+                     closeLock.unlock();
+                  }
+                  break;
+               }
+               // If we couldn't offer an entry check if we were completed concurrently
+               // We have to do in lock to ensure we see updated value properly
+               closeLock.lock();
+               try {
+                  stop = closed;
+               } finally {
+                  closeLock.unlock();
+               }
+            }
+         }
+      }
+   }
+
+   @Override
+   public CacheEntry<K, V> get() throws TimeoutException {
```

Comment:
You're right about parallelism, it does complicate things a bit. For distributed streams, it would be really nice if we could split the parallel iteration into per-node spliterators, and avoid the buffering that happens when you use `IteratorAsSpliterator.split()`, but that might be too tricky, so I'm ok with batching in the spliterators that get returned to the user.

I guess you could use `forEachRemaining()` and throw an exception when the limit is hit (or the predicate matches etc.)? Not sure how much it would complicate the code, though.